### PR TITLE
Data Catalogue

### PR DESCRIPTION
**Pull request by liviajakob**
Changes to include a data catalogue class that reads the metadata json file. and stores all datasets.

### CODE REVIEW

File: glambie/const/data_groups.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+from dataclasses import dataclass
+
+
+@dataclass
+class GlambieDataGroup():
+    name: str
+    long_name: str  # might add some more here later
+
+
+class GlambieDataGroups():
```
Comment (alex-ew) [inline glambie/const/data_groups.py:10]: So far this reads like read-only dictionary. If it doesn't need to be read-only, then just inherit from dict. If it does you can follow one of these examples: https://stackoverflow.com/questions/19022868/how-to-make-dictionary-read-only

But you get all the item look up stuff (and set uniqueness) for free if your dictionary is:
{ "name": GlambieDataGroup}

File: glambie/const/regions.py [hunk 2]
```diff
@@ -10,8 +10,20 @@ class RGIRegion():
...
 
+    def __str__(self):
+        return str(self.rgi_id) + '; ' + self.name + '; ' + self.long_name
 
-class Regions():
+
+class MetaRegions(type):
+
+    def __str__(cls):
+        string = ''
```
Comment (alex-ew) [inline glambie/const/regions.py:20]: Minor one. I think this is a one liner with '\n'.join(cls.regions).
or with list comprehension if you need to force type '\n'.join([str(region) for region in cls.regions]).

File: glambie/data/data_catalogue.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            user_group = ds_dict['user_group']
+            dataset = ChangeTimeseries(data_filepath=fp, region=region, data_group=data_group, user_group=user_group,
+                                       read_data=read_data)
+            self._datasets.append(dataset)
+
+    @staticmethod
+    def from_file(metadata_file_path: str):
+        with open(metadata_file_path) as json_file:
+            meta_data = json.load(json_file)
+            print(type(meta_data))
```
Comment (alex-ew) [inline glambie/data/data_catalogue.py:43]: Should be logged or removed

File: glambie/data/data_catalogue.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            return DataCatalogue(meta_data)
+
+    def as_dataframe(self):
+        metadata_list = [ds.metadata_as_dataframe() for ds in self.datasets]
+        return pd.concat(metadata_list)
+
+    def __len__(self) -> int:
+        return len(self.datasets)
+
+    def get_regions(self) -> list:
```
Comment (alex-ew) [inline glambie/data/data_catalogue.py:53]: The type hint is cool, but it's nicer to see what you'll get a list of. i.e. List[str] etc

File: glambie/const/data_groups.py (hunk 1 above)
Comment (alex-ew) [inline glambie/const/data_groups.py:10]: Think the last comment was lost..(?)

This looks like a readonly dictionary. It probably doesn't need to be readonly, in which case just inherit from dict. If it does need to be readonly, you could follow something like this: 
https://stackoverflow.com/questions/19022868/how-to-make-dictionary-read-only

If you inherit from dict you get this lookup stuff for free

File: glambie/const/data_groups.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    data_groups = [
+        GlambieDataGroup(name='altimetry', long_name='Altimetry'),
+        GlambieDataGroup(name='demdiff', long_name='DEM differencing'),
+        GlambieDataGroup(name='gravimetry', long_name='Gravimetry'),
+        GlambieDataGroup(name='combined', long_name='Combination of methods')
+    ]
+
+    @staticmethod
+    def get(name: str):
+        return GlambieDataGroups.get_data_group_by_name(name)
```
Comment (alex-ew) [inline glambie/const/data_groups.py:22]: If it's a simple duplicate of the below method, just keep one of them?

Comment (alex-ew) [reply glambie/const/data_groups.py:22]: But both disappear with the dict inheritance

**Review by alex-ew — COMMENTED** [review]
Generally looking good - like the tests. I still need to work through some of the files - will pick up in the morning

File: glambie/data/timeseries_data.py [hunk 6]
```diff
@@ -26,21 +30,42 @@ def max_change_value(self) -> float:
         finite_list = np.isfinite(self.changes)
         return np.max(self.changes[finite_list])
 
-    def __init__(self, dates: np.ndarray, area: np.ndarray,
-                 changes: np.ndarray, errors: np.ndarray, rgi_version: int, region_id: int, unit: str,
-                 user: Optional[str] = None, user_group: Optional[str] = None, data_group: Optional[str] = None):
+    def __init__(self, data_group: Optional[GlambieDataGroup] = None, region: Optional[RGIRegion] = None,
```
Comment (alex-ew) [inline glambie/data/timeseries_data.py:33]: We could reuse the dataclass approach here. I had to play with it myself but for optional values you just use something like 
```
from dataclasses import dataclass


@dataclass
class ClassWithDefaults():
    mask: str
    msg: str = None

    def test(self):
        return(self.msg)


cls = ClassWithDefaults(mask="str")

print(cls.mask, cls.msg, cls.test())
```

File: glambie/data/timeseries_data.py [hunk 7]
```diff
@@ -26,21 +30,42 @@ def max_change_value(self) -> float:
...
+        # basin
+        self.region = region
         # unit of timeseries, ie mwe (gravimetry) or m (other)
         self.unit = unit
+        # filepath of data
+        self.data_filepath = data_filepath
+        if read_data and self.data_filepath is not None:
+            self.read_data()
+
+    def read_data(self, data_filepath: Optional[str] = None):
```
Comment (alex-ew) [inline glambie/data/timeseries_data.py:54]: Let's have a chat about this pattern in general. I feel like there's a cleaner way to capture what you're trying to achieve. For my own benefit later: The class has some basic info and then some more information that's populated with ingest_data. But you want the contents to lazy load. It's probably better to have a class state flag self.is_ingesed. And then whenever you need the data have a method ingest_if_necessary() that checks if it's already ingested or not and loads it. Maybe - let's discuss :-)

File: glambie/data/timeseries_data.py [hunk 8]
```diff
@@ -52,13 +77,24 @@ def temporal_resolution(self) -> float:
         resolution = (Decimal(self.max_time) - Decimal(self.min_time)) / len(self)
         return float(resolution)
```
Comment (alex-ew) [inline glambie/data/timeseries_data.py:78]: Should it be a @property like the other dataset info above?

File: glambie/data/data_catalogue.py (hunk 3 above)
Comment (liviajakob) [reply glambie/data/data_catalogue.py:43]: oops that slipped in

File: glambie/data/timeseries.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+                             'errors': self.errors,
+                             'area': self.area
+                             })
+
+
+class Timeseries():
+    """Class containing a data series and corresponding metadata"""
+    is_data_loaded = False
+
+    def __init__(self, region: RGIRegion = None, data_group: GlambieDataGroup = None, data_filepath: str = None,
```
Comment (alex-ew) [inline glambie/data/timeseries.py:59]: A comment in the constructor would be good about the lazy data loading or presence of data as input

File: glambie/data/data_catalogue.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+    @property
+    def base_path(self) -> str:
+        return self._base_path
+
+    def as_dataframe(self) -> pd.DataFrame:
+        metadata_list = [ds.metadata_as_dataframe() for ds in self._datasets]
+        return pd.concat(metadata_list)
+
+    def get_filtered_catalogue(self, region_name: str = None, data_group: str = None,
```
Comment (alex-ew) [inline glambie/data/data_catalogue.py:56]: I think this method in particular would especially benefit from a return type hint and doc strings so it's clear the user gets a new catalogue object back

**Review by alex-ew — CHANGES_REQUESTED** [review]
Nice - things are looking a lot simpler now and the general object hierarchy is clearer. The test coverage is great - it could pass operational standards with a few cosmetic changes. I think we just need a couple more docstrings on key methods and we're done

**Review by alex-ew — APPROVED** [review]
Looking good

**Pull request merged**